include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +393,277 @@ class BidirectionalNode
+    auto Comparator = [&S](auto const &Edge) { return Edge.Neighbor == &S; };
+    return std::find_if(Successors.begin(), Successors.end(), Comparator);
+  }
+  typename EdgeViewContainer::iterator findPredecessor(DerivedType const &P) {
+    auto Comparator = [&P](auto &Edge) { return Edge.Neighbor == &P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
+  }
+  typename EdgeViewContainer::const_iterator
+  findPredecessor(DerivedType const &P) const {
+    auto Comparator = [&P](auto const &Edge) { return Edge.Neighbor == &P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
+  }
+
+public:
+  bool hasSuccessor(DerivedType const &S) const {
+    return findSuccessor(S) != Successors.end();
+  }
+  bool hasPredecessor(DerivedType const &S) const {
+    return findPredecessor(S) != Predecessors.end();
+  }
+
+public:
+  size_t successorCount() const { return Successors.size(); }
+  size_t predecessorCount() const { return Predecessors.size(); }
+
+  bool hasSuccessors() const { return Successors.size() != 0; }
+  bool hasPredecessors() const { return Predecessors.size() != 0; }
+
+public:
+  // Maybe this overload should be `protected`. But it's faster than the
+  // alternative, so I'm hesitant.
+  typename EdgeOwnerContainer::iterator
+  removeSuccessor(typename EdgeOwnerContainer::const_iterator SuccessorIt) {
+    // Maybe we should do some checks as to whether `SuccessorIt` is valid.
+
+    auto *Successor = SuccessorIt->Neighbor;
+    auto PredecessorIt = Successor->findPredecessor(*this);
+    revng_assert(PredecessorIt != Successor->Predecessors.end(),
+                 "Half of an edge is missing, graph layout is broken.");
+    Successor->Predecessors.erase(PredecessorIt);
+    return Successors.erase(SuccessorIt);
+  }
+  auto removeSuccessor(DerivedType const &S) {
+    return removeSuccessor(findSuccessor(S));
+  }
+
+  // Maybe this overload should be `protected`. But it's faster than the
+  // alternative, so I'm hesitant.
+  typename EdgeViewContainer::iterator
+  removePredecessor(typename EdgeViewContainer::const_iterator PredecessorIt) {
+    // Maybe we should do some checks as to whether `PredecessorIt` is valid.
+
+    auto *Predecessor = PredecessorIt->Neighbor;
+    auto SuccessorIt = Predecessor->findSuccessor(*this);
+    revng_assert(SuccessorIt != Predecessor->Successors.end(),
+                 "Half of an edge is missing, graph layout is broken.");
+    Predecessor->Successors.erase(SuccessorIt);
+    return Predecessors.erase(PredecessorIt);
+  }
+  auto removePredecessor(DerivedType const &P) {
+    return removePredecessor(findPredecessor(P));
+  }
+
+public:
+  NewNode &disconnect() {
+    for (auto It = Successors.begin(); It != Successors.end(); ++It)
+      It = removeSuccessor(It);
+    for (auto It = Predecessors.begin(); It != Predecessors.end(); ++It)
+      It = removePredecessor(It);
+    return *this;
+  }
```

Comment:
Is this so common to implement it for the user?

Reply:
Not really, I guess. Here it's more so that `Graph::removeNode` is not as cluttered. But yeah, it can be clearly moved there.

Reply:
But I'm somewhat hesitant on placing a complex-ish stuff inside this `if constexpr` condition, maybe we just make this method protected, if you don't want to show it to the user, but then we'd need to declare the graph this node's friend and stuff.

Reply:
So what should I do with it? Should I move it? Or is it Ok the way it is?

Reply:
It's OK as it is